sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -255,6 +257,81 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
   return shutdown_redact_policy_inline(raw_state);
 }
 
+enum {
+  /**
+   * Length of the error message prefix.
+   */
+  kMsgPrefixLen = 10,
+  /**
+   * Length of the hexadecimal representation.
+   */
+  kHexStrLen = 8,
+  /**
+   * Total message length.
+   *
+   * This includes the ": 0x" before the hex string '\n' at the end of the
+   * message.
+   */
+  kTotalMsgLen = kMsgPrefixLen + kHexStrLen + 5,
+  /**
+   * Base address of UART.
+   */
+  kUartBase = TOP_EARLGREY_UART0_BASE_ADDR,
+  /**
+   * UART TX FIFO size.
+   */
+  kUartFifoSize = 32,
+};
+static_assert(kTotalMsgLen <= kUartFifoSize,
+              "Total message length must be less than TX FIFO size.");
+
+/**
+ * Prints a fixed-length (`kMsgPrefixLen`) message prefix followed by the
+ * hexadecimal representation of a `uint32_t`.
+ *
+ * This function prints ": 0x" before printing the hex string and prints a '\n'
+ * at the end of the message. This function also blocks until UART transmission
+ * ends.
+ *
+ * @param prefix Prefix to print.
+ * @param val Integer to print.
+ */
+static ALWAYS_INLINE void shutdown_print(const char *prefix, uint32_t val) {
+  for (size_t i = 0; i < kMsgPrefixLen; ++i) {
+    // `kMsgPrefixLen` is laundered so that it is loaded to a register at every
+    // iteration.
+    if (i >= launder32(kMsgPrefixLen)) {
+      return;
+    }
+    abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix[i]);
+  }
+
+  // Print ": 0x" before the hex string.
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, ':');
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, ' ');
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, '0');
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, 'x');
+
+  const char kHexTable[16] = "0123456789abcdef";
+  for (size_t i = 0; i < kHexStrLen; ++i) {
+    uint8_t nibble = bitfield_field32_read(
+        val, (bitfield_field32_t){.mask = 0xf, .index = (7 - i) * 4});
+    // `kHexStrLen` is laundered so that it is loaded to a register at every
+    // iteration.
+    if (i >= launder32(kHexStrLen)) {
+      return;
+    }
+    abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, kHexTable[nibble]);
+  }
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, '\n');
+
+  // Wait until UART TX is idle.
+  uint32_t status_reg;
+  do {
+    status_reg = abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET);
+  } while (!bitfield_bit32_read(status_reg, UART_STATUS_TXIDLE_BIT));
```

Comment:
Checked the UART HWIP spec again: TXIDLE is true when either the FIFO is non-empty or there is a transmission. Thus this loop blocks until all characters are transmitted (may never happen if UART is not transmitting for any reason). Enabling transmission unconditionally was an attempt at addressing this but a limit on the number of iterations would be better. Determining this constant requires some work because CPU clock frequency and baud rate depends on the device and they are link-time constants. If we just define the constant here, we end up with multiplication and division operations so I'll trying adding a constant like `kUartNCOValue` and push an update shortly.